pkg/source/hdfsprotocol/hdfs_source_client.go

```diff
@@ -17,52 +17,232 @@
-	panic("implement me")
+func (h *hdfsSourceClient) GetContentLength(ctx context.Context, url string, header source.RequestHeader) (int64, error) {
+	hdfsClient, path, err := h.getHDFSClientAndPath(url)
+	if err != nil {
+		return -1, err
+	}
+	info, err := hdfsClient.Stat(path)
+	if err != nil {
+		return -1, err
+	}
+	return info.Size(), nil
+}
+
+func (h *hdfsSourceClient) IsSupportRange(ctx context.Context, url string, header source.RequestHeader) (bool, error) {
+	return true, nil
+}
+
+func (h *hdfsSourceClient) IsExpired(ctx context.Context, url string, header source.RequestHeader, expireInfo map[string]string) (bool, error) {
+	lastModified := expireInfo[headers.LastModified]
+	//eTag := expireInfo[headers.ETag]
+	if lastModified == "" {
+		return true, nil
+	}
+
+	hdfsClient, path, err := h.getHDFSClientAndPath(url)
+	if err != nil {
+		return false, err
+	}
+
+	info, err := hdfsClient.Stat(path)
+	if err != nil {
+		return false, err
+	}
+	t, err := time.ParseInLocation(layout, lastModified, time.Local)
+	if err != nil {
+		return false, err
+	}
+	return info.ModTime().Format(layout) != t.Format(layout), nil
+}
+
+func (h *hdfsSourceClient) Download(ctx context.Context, url string, header source.RequestHeader) (io.ReadCloser, error) {
+	hdfsClient, path, err := h.getHDFSClientAndPath(url)
+	if err != nil {
+		return nil, err
+	}
+	hdfsFile, err := hdfsClient.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer hdfsFile.Close()
+
+	data, err := ioutil.ReadAll(hdfsFile)
+	if err != nil {
+		return nil, err
+	}
+	return ioutil.NopCloser(bytes.NewBuffer(data)), nil
 }
 
-func (h hdfsSourceClient) IsSupportRange(ctx context.Context, url string, header source.RequestHeader) (bool, error) {
-	panic("implement me")
+func (h *hdfsSourceClient) DownloadWithResponseHeader(ctx context.Context, url string, header source.RequestHeader) (io.ReadCloser, source.ResponseHeader, error) {
+
+	breakPoint := header.Get(cdn.RangeHeaderName)
+
+	if breakPoint == "" {
+		// default start break 0
+		breakPoint = "0"
+	}
+
+	breakPointLSize, err := strconv.ParseInt(breakPoint, 0, 64)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	hdfsClient, path, err := h.getHDFSClientAndPath(url)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	hdfsFile, err := hdfsClient.Open(path)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer hdfsFile.Close()
+
+	fileInfo := hdfsFile.Stat()
+
+	if breakPointLSize > fileInfo.Size() {
+		return nil, nil, errors.Errorf("hdfs Range more then file size,range start %d, file size %d", breakPointLSize, fileInfo.Size())
+	}
+
+	_, err = hdfsFile.Seek(breakPointLSize, 0)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	expectReadSize := fileInfo.Size() - breakPointLSize
+	data := make([]byte, expectReadSize)
+	readLength, err := hdfsFile.Read(data)
+	if err != nil || int64(readLength) != expectReadSize {
+		return nil, nil, err
+	}
+
+	return ioutil.NopCloser(bytes.NewBuffer(data)), source.ResponseHeader{
```

Comment:
After seek, wrap hdfsFile into io.LimitedReader

Reply:
For range, seek and the wrap hdfsFile into io.LimitedReader.
Add a new type to handle close.
```
type x struct {
  io.LimitedReader
  c io.Closer
}

func (*x) WriteTo() {
}

func (*x) Close() {
}
```

Reply:
done